src/Rules/PHPUnit/DataProviderHelper.php

```diff
@@ -0,0 +1,111 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\PHPUnit;
+
+use PHPStan\Analyser\Scope;
+use PHPStan\PhpDoc\ResolvedPhpDocBlock;
+use PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocTagNode;
+use PHPStan\Reflection\MissingMethodFromReflectionException;
+use PHPStan\Rules\RuleError;
+use PHPStan\Rules\RuleErrorBuilder;
+use function array_merge;
+use function preg_match;
+use function sprintf;
+use function trim;
+
+class DataProviderHelper
+{
+
+	/**
+	 * @return array<PhpDocTagNode>
+	 */
+	public function getDataProviderAnnotations(?ResolvedPhpDocBlock $phpDoc): array
+	{
+		if ($phpDoc === null) {
+			return [];
+		}
+
+		$phpDocNodes = $phpDoc->getPhpDocNodes();
+
+		$annotations = [];
+
+		foreach ($phpDocNodes as $docNode) {
+			$annotations = array_merge(
+				$annotations,
+				$docNode->getTagsByName('@dataProvider')
+			);
+		}
+
+		return $annotations;
+	}
+
+	/**
+	 * @return RuleError[] errors
+	 */
+	public function processDataProvider(
+		Scope $scope,
+		PhpDocTagNode $phpDocTag
+	): array
+	{
+		$dataProviderName = $this->getDataProviderName($phpDocTag);
+		if ($dataProviderName === null) {
+			// Missing name is already handled in NoMissingSpaceInMethodAnnotationRule
+			return [];
+		}
+
+		$classReflection = $scope->getClassReflection();
+		if ($classReflection === null) {
+			// Should not happen
+			return [];
+		}
+
+		try {
+			$dataProviderMethodReflection = $classReflection->getMethod($dataProviderName, $scope);
+		} catch (MissingMethodFromReflectionException $missingMethodFromReflectionException) {
+			$error = RuleErrorBuilder::message(sprintf(
+				'@dataProvider %s related method not found.',
+				$dataProviderName
+			))->build();
+
+			return [$error];
+		}
+
+		$errors = [];
+
+		if ($dataProviderName !== $dataProviderMethodReflection->getName()) {
+			$errors[] = RuleErrorBuilder::message(sprintf(
+				'@dataProvider %s related method is used with incorrect case: %s.',
+				$dataProviderName,
+				$dataProviderMethodReflection->getName()
+			))->build();
+		}
+
+		if (!$dataProviderMethodReflection->isPublic()) {
+			$errors[] = RuleErrorBuilder::message(sprintf(
+				'@dataProvider %s related method must be public.',
+				$dataProviderName
+			))->build();
+		}
+
+		if (!$dataProviderMethodReflection->isStatic()) {
+			$errors[] = RuleErrorBuilder::message(sprintf(
+				'@dataProvider %s related method must be static.',
```

Comment:
This is too strict. Current version of PHPUnit doesn't require this, and for example all dataProviders in PHPStan are non-static.

Reply:
Well, that was the idea behind #148, to ease the migrations to the next version.

Would you be ok to make this configurable?

Reply:
Removed for now